source/agent/addons/quic/QuicTransportStream.cc

```diff
@@ -0,0 +1,225 @@
+    obj->Wrap(info.This());
+    uv_async_init(uv_default_loop(), &obj->m_asyncOnContentSessionId, &QuicTransportStream::onContentSessionId);
+    uv_async_init(uv_default_loop(), &obj->m_asyncOnData, &QuicTransportStream::onData);
+    info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(QuicTransportStream::write){
+    if(info.Length()<2){
+        Nan::ThrowTypeError("Data is not provided.");
+        return;
+    }
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info.Holder());
+    uint8_t* buffer=(uint8_t*)node::Buffer::Data(info[0]->ToObject());
+    unsigned int length = info[1]->Uint32Value();
+    obj->m_stream->Write(buffer, length);
+    info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(QuicTransportStream::addDestination)
+{
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info.Holder());
+    if (info.Length() != 1) {
+        Nan::ThrowTypeError("Invalid argument length for addDestination.");
+        return;
+    }
+    // TODO: Check if info[0] is an Nan wrapped object.
+    auto framePtr = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info[0]->ToObject());
+    // void* ptr = info[0]->ToObject()->GetAlignedPointerFromInternalField(0);
+    // auto framePtr=static_cast<owt_base::FrameDestination*>(ptr);
+    obj->addDataDestination(framePtr);
+    obj->m_isPiped = true;
+}
+
+NAN_METHOD(QuicTransportStream::removeDestination)
+{
+}
+
+v8::Local<v8::Object> QuicTransportStream::newInstance(owt::quic::QuicTransportStreamInterface* stream)
+{
+    Local<Object> streamObject = Nan::NewInstance(Nan::New(QuicTransportStream::s_constructor)).ToLocalChecked();
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(streamObject);
+    obj->m_stream = stream;
+    obj->MaybeReadContentSessionId();
+    return streamObject;
+}
+
+void QuicTransportStream::MaybeReadContentSessionId()
+{
+    if (!m_receivedContentSessionId) {
+        // Match to a content session.
+        if (m_stream->ReadableBytes() > 0 && m_stream->ReadableBytes() < uuidSizeInByte) {
+            ELOG_ERROR("No enough data to get content session ID.");
+            m_stream->Close();
+            return;
+        }
+        uint8_t* data = new uint8_t[uuidSizeInByte];
+        m_stream->Read(data, uuidSizeInByte);
+        m_contentSessionId = std::vector<uint8_t>(data, data + uuidSizeInByte);
+        m_receivedContentSessionId = true;
+        m_asyncOnContentSessionId.data = this;
+        uv_async_send(&m_asyncOnContentSessionId);
+        if (m_stream->ReadableBytes() > 0) {
+            SignalOnData();
+        }
+    }
+}
+
+NAUV_WORK_CB(QuicTransportStream::onData){
+    Nan::HandleScope scope;
+    QuicTransportStream* obj = reinterpret_cast<QuicTransportStream*>(async->data);
+    if (obj == nullptr) {
+        return;
+    }
+    Nan::MaybeLocal<v8::Value> onEvent = Nan::Get(obj->handle(), Nan::New<v8::String>("ondata").ToLocalChecked());
```

Comment:
you need to check m_receivedContentSessionId here and handle accordingly.

Reply:
No need to do so because it is checked before firing `onData` event. See `OnCanRead`.